frontend/src/wasm-communication/messages.ts

```diff
@@ -1188,13 +1190,46 @@ export type WidgetLayout = {
 	layout: LayoutGroup[];
 };
 
+export type WidgetDiffUpdate = {
+	layoutTarget: unknown;
+	diff: WidgetDiff[];
+};
+
+type WidgetDiff = { path: number[]; new: LayoutGroup[] | LayoutGroup | Widget | MenuBarEntry[] };
+
 export function defaultWidgetLayout(): WidgetLayout {
 	return {
 		layoutTarget: undefined,
 		layout: [],
 	};
 }
 
+// Updates a widget layout based on a list of updates, returning the new layout
+export function patchWidgetLayout(layout: WidgetLayout, updates: WidgetDiffUpdate): void {
+	layout.layoutTarget = updates.layoutTarget;
+
+	updates.diff.forEach((update) => {
+		// Find the object where the diff applies to
+		let targetLayout = layout.layout as LayoutGroup[] | LayoutGroup | Widget | MenuBarEntry[] | MenuBarEntry | undefined;
+		update.path.forEach((index) => {
+			if (!targetLayout) return;
+			if ("columnWidgets" in targetLayout) targetLayout = targetLayout.columnWidgets[index];
+			else if ("rowWidgets" in targetLayout) targetLayout = targetLayout.rowWidgets[index];
+			else if ("layout" in targetLayout) targetLayout = targetLayout.layout[index];
+			// eslint-disable-next-line no-console
+			else if (targetLayout instanceof Widget) console.error("Tried to index widget");
+			else if ("action" in targetLayout) targetLayout = targetLayout.children ? targetLayout.children[index] : undefined;
+			else targetLayout = targetLayout[index];
+		});
+
+		// Some odd code required to update the objects (assigning things doesn't work?)
+		if (targetLayout !== undefined && "length" in targetLayout) targetLayout.length = 0;
```

Comment:
These three lines seem sketchy. Do you mind adding some comments on why exactly they are needed?

Reply:
> Some odd code required to update the objects (assigning things doesn't work?)

Already commented.

Reply:
I don't really know why javascript likes this to be honest. Something to do with how the implicit pointers are done.

Reply:
Some comments on what you are expecting the shapes of the objects would be helpful or maybe some links to stack overflow. I have never seen something needed like this in all my web experiance

Reply:
It is probably possible to avoid this but I don't know enough js to do so.